virtualfish/virtual.fish

```diff
@@ -129,13 +169,92 @@ function __vf_new --description "Create a new virtualenv"
     end
 
     emit virtualenv_will_create
-    set envname $argv[-1]
-    set -e argv[-1]
-    if set -q VIRTUALFISH_DEFAULT_PYTHON
-        set argv "--python" $VIRTUALFISH_DEFAULT_PYTHON $argv
+    argparse -n "vf new" -x q,v,d --ignore-unknown "h/help" "q/quiet" "v/verbose" "d/debug" "p/python=" -- $argv
+
+    if set -q _flag_help
+        echo "Purpose: Creates a new virtual environment"
+        echo "Usage: "(set_color green)"vf new "(set_color -di)"[-p <python-version>] [<virtualenv-flags>]"(set_color normal)(set_color green)" <virtualenv-name>"(set_color normal)
+        echo
+        echo "Examples:"
+        echo
+        echo (set_color green)"vf new -p /usr/local/bin/python3 yourproject"(set_color normal)
+        echo (set_color green)"vf new -p python3.8 --system-site-packages yourproject"(set_color normal)
+        echo
+        echo "To see available "(set_color blue)"Virtualenv"(set_color normal)" option flags, run: "(set_color green)"virtualenv --help"(set_color normal)
+        return 0
+    end
+
+    # Unpack Virtualenv args: first flags that need values, then Boolean flags
+    # When done, $argv should only contain the new virtual environment name.
+    set -l flags_with_args "--app-data" "--discovery" "--creator" "--seeder" "--activators" "--extra-search-dir" "--pip" "--setuptools" "--wheel" "--prompt"
+    for arg in $argv
+        # If arg starts with a hyphen…
+        if test (string sub -l 1 -- $arg) = "-"
+            # Retrieve its index number
+            set -l idx (contains -i -- $arg $argv)
```

Comment:
This breaks if an option is ever given twice - `contains -i` will print the first match (yes, there is an issue open about adding a "--all-indices").

I suggest either using `for i in (seq (count $argv))` or `while set -q argv[1]` instead of looping over the elements if you need "the next".

Reply:
I don’t believe the `virtualenv` command supports duplicate option flags. But just in case, I tried with the existing for-loop, and unless I’m misunderstanding something, it seems duplicate option flags are handled appropriately:

    ➤ vf new --debug --clear --clear foo
    Creating foo via /usr/local/bin/python …
    Virtualenv args: --python /usr/local/bin/python --clear --clear
    Extra args: foo